File: src/policy/sine_arm_policy.cpp
#include "policy/sine_arm_policy.hpp"

#include "policy/arm_demo_common.hpp"

#include <cmath>
#include <stdexcept>

namespace policy_runner {
namespace {

constexpr float kElbowBase = 0.0f;
constexpr float kElbowAmp = 0.4f;
constexpr float kFreqHz = 0.5f;

}  // namespace

SineArmPolicy::SineArmPolicy(float control_dt) : control_dt_(control_dt) {}

std::string SineArmPolicy::name() const { return "sine_arm"; }

std::size_t SineArmPolicy::input_dim() const { return kB1JointCount; }

std::vector<int> SineArmPolicy::controlled_joints() const {
  return {arm_demo::kLeftElbow, arm_demo::kRightElbow};
}

Observation SineArmPolicy::BuildObservation(
    const RobotState& state,
    const std::vector<float>& /*command*/) const {
  if (state.q.size() != kB1JointCount) {
    throw std::runtime_error("SineArmPolicy: RobotState.q size mismatch");
  }
  Observation obs;
  obs.data = state.q;
  return obs;
}

Action SineArmPolicy::Infer(const Observation& obs) {
  if (obs.data.size() != input_dim()) {
    throw std::runtime_error("SineArmPolicy: observation size mismatch");
  }

  const float phase = 2.f * 3.14159265f * kFreqHz * time_s_;
  const float elbow_q = kElbowBase + kElbowAmp * std::sin(phase);

  Action action;
  action.joint_cmds.push_back(
      arm_demo::MakeElbowCmd(arm_demo::kLeftElbow, elbow_q));
  action.joint_cmds.push_back(
      arm_demo::MakeElbowCmd(arm_demo::kRightElbow, elbow_q));

  time_s_ += control_dt_;
  return action;
}

void SineArmPolicy::Reset() { time_s_ = 0.f; }

}  // namespace policy_runner

File: python-implementation/policy_runner/policy/sine_arm_policy.py
"""Hard-coded sine-wave elbow policy (mirrors C++ SineArmPolicy)."""

from __future__ import annotations

import math
from typing import List, Sequence

from policy_runner.policy.arm_demo_common import (
    LEFT_ELBOW,
    RIGHT_ELBOW,
    make_elbow_cmd,
)
from policy_runner.policy.base import Policy
from policy_runner.types import (
    BASE_OBS_DIM,
    Action,
    Observation,
    RobotState,
    pack_observation,
)

_ELBOW_BASE = 0.0
_ELBOW_AMP = 0.4
_FREQ_HZ = 0.5


class SineArmPolicy(Policy):
    """Elbows (5, 9) follow a sine wave. Sparse output — other joints weight=0."""

    def __init__(self, control_dt: float = 0.02) -> None:
        self._control_dt = control_dt
        self._time_s = 0.0

    def name(self) -> str:
        return "sine_arm"

    def input_dim(self) -> int:
        return BASE_OBS_DIM

    def controlled_joints(self) -> List[int]:
        return [LEFT_ELBOW, RIGHT_ELBOW]

    def build_observation(
        self, state: RobotState, command: Sequence[float]
    ) -> Observation:
        return pack_observation(state, command)

    def infer(self, obs: Observation) -> Action:
        self.assert_frame_observation(obs)

        phase = 2.0 * math.pi * _FREQ_HZ * self._time_s
        elbow_q = _ELBOW_BASE + _ELBOW_AMP * math.sin(phase)

        self._time_s += self._control_dt
        return Action(
            joint_cmds=[
                make_elbow_cmd(LEFT_ELBOW, elbow_q),
                make_elbow_cmd(RIGHT_ELBOW, elbow_q),
            ]
        )

    def reset(self) -> None:
        self._time_s = 0.0
